File: src/abstraction/v1-8-9/java/net/eduard/api/lib/abstraction/PetsAPI.java
package net.eduard.api.lib.abstraction;

import net.minecraft.server.v1_8_R3.*;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.craftbukkit.v1_8_R3.entity.CraftLivingEntity;
import org.bukkit.craftbukkit.v1_8_R3.util.UnsafeList;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Player;

import java.lang.reflect.Field;
import java.util.UUID;

public class PetsAPI {

	 private static Field gsa;
	    private static Field goalSelector;
	    private static Field targetSelector;
	    static {
	        try {
	            gsa = PathfinderGoalSelector.class.getDeclaredField("b");
	            gsa.setAccessible(true);
	            goalSelector = EntityInsentient.class.getDeclaredField("goalSelector");
	            goalSelector.setAccessible(true);
	            targetSelector = EntityInsentient.class.getDeclaredField("targetSelector");
	            targetSelector.setAccessible(true);
	        } catch (Exception e) {
	            e.printStackTrace();
	        }
	    }
	    @SuppressWarnings("deprecation")
	    public static void makePet(LivingEntity e, UUID toFollow) {
	        try {
	            Object nms_entity = ((CraftLivingEntity) e).getHandle();
	            if (nms_entity instanceof EntityInsentient) {
	                PathfinderGoalSelector goal = (PathfinderGoalSelector) goalSelector.get(nms_entity);
	                PathfinderGoalSelector target = (PathfinderGoalSelector) targetSelector.get(nms_entity);
	                gsa.set(goal, new UnsafeList<Object>());
	                gsa.set(target, new UnsafeList<Object>());
	                goal.a(0, new PathfinderGoalFloat((EntityInsentient) nms_entity));
	                goal.a(1, new PathfinderGoalWalktoTile((EntityInsentient) nms_entity, toFollow));
	            } else {
	                throw new IllegalArgumentException(e.getType().getName() + " is not an instance of an EntityInsentient.");
	            }
	        } catch (Exception ex) {
	            ex.printStackTrace();
	        }
	    }
	    public static class PathfinderGoalWalktoTile extends PathfinderGoal {
	        private final EntityInsentient entity;
	        private PathEntity path;
	        private final UUID uuid;

	        public PathfinderGoalWalktoTile(EntityInsentient entitycreature, UUID playerUUID) {
	            this.entity = entitycreature;
	            this.uuid = playerUUID;

	        }
	        @Override
	        public boolean a() {
	        	Player alvo = Bukkit.getPlayer(uuid);
	            if (Bukkit.getPlayer(uuid) == null) {
	                return path != null;
	            }
	            Location targetLocation = alvo.getLocation();
				//boolean flag = this.entity.getNavigation().m();
	            this.path = this.entity.getNavigation().a(targetLocation.getX() + 1, targetLocation.getY(), targetLocation.getZ() + 1);

	            if (this.path != null) {
	                this.c();
	            }
	            return this.path != null;
	        }
	        @Override
	        public void c() {
	        	this.entity.getNavigation().a(this.path, 1D);
	        }
	    }		
}

File: MineNMS-1_8_9/src/main/java/PetsAPI.kt
package net.eduard.api.lib.abstraction

import java.util.UUID
import org.bukkit.craftbukkit.v1_8_R3.entity.CraftLivingEntity
import net.minecraft.server.v1_8_R3.EntityInsentient
import net.minecraft.server.v1_8_R3.PathfinderGoalSelector
import org.bukkit.craftbukkit.v1_8_R3.util.UnsafeList
import net.minecraft.server.v1_8_R3.PathfinderGoalFloat
import java.lang.IllegalArgumentException
import net.minecraft.server.v1_8_R3.PathfinderGoal
import net.minecraft.server.v1_8_R3.PathEntity
import org.bukkit.Bukkit
import org.bukkit.entity.LivingEntity
import java.lang.Exception
import java.lang.reflect.Field

object PetsAPI {
    private lateinit var pathFinderListField: Field
    private lateinit var goalSelectorField: Field
    private lateinit var targetSelectorField: Field
    fun makePetFollowPlayer(petEntity: LivingEntity, playerUUID: UUID) {
        try {
            val nmsEntity: Any = (petEntity as CraftLivingEntity).handle
            if (nmsEntity is EntityInsentient) {
                val goal = goalSelectorField[nmsEntity] as PathfinderGoalSelector
                val target = targetSelectorField[nmsEntity] as PathfinderGoalSelector
                pathFinderListField[goal] = UnsafeList<Any>()
                pathFinderListField[target] = UnsafeList<Any>()
                goal.a(0, PathfinderGoalFloat(nmsEntity))
                goal.a(1, PathFinderFollowPlayer(nmsEntity, playerUUID))
            } else {
                throw IllegalArgumentException(
                    petEntity.getType().getName() + " is not an instance of an EntityInsentient."
                )
            }
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }

    class PathFinderFollowPlayer(
        val entity: EntityInsentient,
        val playerID: UUID
    ) : PathfinderGoal() {
        var path: PathEntity? = null
        override fun a(): Boolean {
            val target = Bukkit.getPlayer(playerID) ?: return path != null
            val targetLocation = target.location
            //boolean flag = this.entity.getNavigation().m();
            path = entity.navigation.a(targetLocation.x + 1, targetLocation.y, targetLocation.z + 1)
            if (path != null) {
                c()
            }
            return path != null
        }

        override fun e() {
        }

        override fun d() {
        }

        override fun i(): Boolean {
            return true
        }

        override fun c() {
            entity.navigation.a(path, 1.0)
        }
    }

    init {
        try {
            pathFinderListField = PathfinderGoalSelector::class.java.getDeclaredField("b")
            pathFinderListField.isAccessible = true
            goalSelectorField = EntityInsentient::class.java.getDeclaredField("goalSelector")
            goalSelectorField.isAccessible = true
            targetSelectorField = EntityInsentient::class.java.getDeclaredField("targetSelector")
            targetSelectorField.isAccessible = true
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}